language/stdlib/modules/libra_system.mvir

```diff
@@ -392,49 +525,80 @@ module LibraSystem {
             }
             // ensures removals.length == 0
             // ensures validators.length == old(validators).length - old(removals).length
-            changed = true;
+            // ensures discovery_set.length == old(discovery_set).length - old(removals).length
+            validator_set_changed = true;
+            discovery_set_changed = true;
         }
 
-        // (3) Check for key rotation
+        // (3) Check for key rotations in the validator set
         i = 0;
-        len = Vector.length<Self.ValidatorInfo>(freeze(copy(vector_ref)));
+        len = Vector.length<Self.ValidatorInfo>(freeze(copy(validators_vec_ref)));
         // assume(len > 0), since an empty validator set is nonsensical
-        info_ref = Vector.borrow_mut<Self.ValidatorInfo>(copy(vector_ref), copy(i));
+        validator_info_ref = Vector.borrow_mut<Self.ValidatorInfo>(copy(validators_vec_ref), copy(i));
         // check if each validator has rotated their keys, copy their new info and note the change
         // if so.
         loop {
-            if (Self.copy_validator_info(move(info_ref))) {
-                changed = true;
+            if (Self.copy_validator_info(move(validator_info_ref))) {
+                validator_set_changed = true;
             }
 
             i = move(i) + 1;
             if (copy(i) >= copy(len)) {
                 break;
             }
-            info_ref = Vector.borrow_mut<Self.ValidatorInfo>(copy(vector_ref), copy(i));
+            validator_info_ref = Vector.borrow_mut<Self.ValidatorInfo>(copy(validators_vec_ref), copy(i));
         }
 
-        // (4) emit changed event if appropriate
-        if (move(changed)) {
+        // (5) Check for changes in the discovery set
+        i = 0;
+        len = Vector.length<Self.DiscoveryInfo>(freeze(copy(discovery_vec_ref)));
+        // assume(len > 0), since an empty validator set is nonsensical
+        discovery_info_ref = Vector.borrow_mut<Self.DiscoveryInfo>(copy(discovery_vec_ref), copy(i));
+        // check if each validator has changed their discovery info, copy their new info, and
+        // note the change if so.
+        loop {
+            if (Self.copy_discovery_info(move(discovery_info_ref))) {
+                discovery_set_changed = true;
+            }
+
+            i = move(i) + 1;
+            if (copy(i) >= copy(len)) {
+                break;
+            }
+            discovery_info_ref = Vector.borrow_mut<Self.DiscoveryInfo>(copy(discovery_vec_ref), copy(i));
+        }
+
+        // (4) Emit validator set changed event if appropriate
+        if (move(validator_set_changed)) {
             LibraAccount.emit_event<Self.ValidatorSetChangeEvent>(
-                &mut move(validator_set_ref).change_events,
+                // TODO(philiphayes): @sam, is this ok to use copy vs move?
```

Comment:
It is completely fine, but slightly cheaper to use a move. The VM will have to make an extra copy of the reference value in `validator_set_ref` if you use a copy here.

The Move source language will just let you write the variable name and figure out whether to emit a copy or move, so not worth worrying about too much.

Reply:
ahh I used `copy` in this last block since compilation failed with JOIN_ERROR if I used `move` for some reason...

like this:

```rust
// ERROR: JOIN_FAILURE

if (move(validator_set_changed)) {
    LibraAccount.emit_event<Self.ValidatorSetChangeEvent>(
        &mut move(validator_set_ref).change_events,
        ValidatorSetChangeEvent {
            new_validator_set: *move(validators_vec_ref),
        },
    );
}

return;

// SUCCESS

if (move(validator_set_changed)) {
    LibraAccount.emit_event<Self.ValidatorSetChangeEvent>(
        &mut copy(validator_set_ref).change_events,
        ValidatorSetChangeEvent {
            new_validator_set: *copy(validators_vec_ref),
        },
    );
}

return;
```

Reply:
Ah, I see why now. The memory pointed to by `validator_set_ref` and `validators_vec_ref` has some overlap, so the bytecode verifier forces you to do the copy. 

If you did something like
```
  new_validator_set =  *move(validators_vec_ref);
  LibraAccount.emit_event<Self.ValidatorSetChangeEvent>(
        &mut move(validator_set_ref).change_events,
        ValidatorSetChangeEvent {
            new_validator_set: move(new_validator_set)
        },
    );
```

, the bytecode verifier would also be happy. But the code looks better to me as-is.